File: src/main/java/com/tkr/desk/ExposureAggregator.java
package com.tkr.desk;

import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.BatchWireFrame;
import com.tkr.types.WireTypes.Status;
import com.tkr.types.WireTypes.WireBatchRecord;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Rolls up gross and net exposure by symbol and account for an allocation batch.
 */
public final class ExposureAggregator {

    public static final class ExposureKey {
        public final int accountId;
        public final int symbolId;

        public ExposureKey(int accountId, int symbolId) {
            this.accountId = accountId;
            this.symbolId = symbolId;
        }

        @Override
        public int hashCode() {
            return 31 * accountId + symbolId;
        }

        @Override
        public boolean equals(Object o) {
            if (!(o instanceof ExposureKey other)) {
                return false;
            }
            return accountId == other.accountId && symbolId == other.symbolId;
        }
    }

    public static final class ExposureSlice {
        public ExposureKey key;
        public long grossNotionalCents;
        public long netQtyMilli;
        public int tradeCount;
        public int maxSingleTradeQty;
        public long weightedAvgPriceTick;
    }

    public static final class ExposureReport {
        public Status status = Status.OK;
        public List<ExposureSlice> slices = new ArrayList<>();
        public long totalGrossNotionalCents;
        public int distinctAccounts;
        public int distinctSymbols;
    }

    private final int deskId;
    private final boolean includePartialFills;

    public ExposureAggregator(int deskId, boolean includePartialFills) {
        this.deskId = deskId;
        this.includePartialFills = includePartialFills;
    }

    public ExposureReport aggregate(BatchWireFrame frame) {
        ExposureReport report = new ExposureReport();
        if (frame == null || frame.records == null) {
            report.status = Status.BOUNDS_ERROR;
            return report;
        }

        Map<ExposureKey, ExposureSlice> map = new HashMap<>();
        for (WireBatchRecord rec : frame.records) {
            if (rec.qtyMilli <= 0) {
                continue;
            }
            if (!includePartialFills
                    && (rec.flags & WireTypes.BATCH_FLAG_PARTIAL_FILL) != 0) {
                continue;
            }

            ExposureKey key = new ExposureKey(rec.accountId, rec.symbolId);
            ExposureSlice slice = map.computeIfAbsent(key, k -> {
                ExposureSlice s = new ExposureSlice();
                s.key = k;
                return s;
            });

            long notional = (long) rec.qtyMilli * (long) rec.priceTick / 1000L;
            slice.grossNotionalCents += notional;
            slice.netQtyMilli += rec.qtyMilli;
            slice.tradeCount++;
            slice.maxSingleTradeQty = Math.max(slice.maxSingleTradeQty, rec.qtyMilli);
            slice.weightedAvgPriceTick =
                    weightedAverage(slice.weightedAvgPriceTick, slice.tradeCount - 1, rec.priceTick);
            report.totalGrossNotionalCents += notional;
        }

        report.slices.addAll(map.values());
        report.distinctAccounts = countDistinctAccounts(report.slices);
        report.distinctSymbols = countDistinctSymbols(report.slices);
        applyConcentrationPenalty(report, deskId);
        return report;
    }

    public ExposureReport mergeReports(List<ExposureReport> parts) {
        ExposureReport merged = new ExposureReport();
        if (parts == null || parts.isEmpty()) {
            return merged;
        }
        Map<ExposureKey, ExposureSlice> map = new HashMap<>();
        for (ExposureReport part : parts) {
            if (part.status != Status.OK) {
                merged.status = part.status;
                return merged;
            }
            for (ExposureSlice slice : part.slices) {
                ExposureSlice target = map.computeIfAbsent(slice.key, k -> {
                    ExposureSlice s = new ExposureSlice();
                    s.key = k;
                    return s;
                });
                target.grossNotionalCents += slice.grossNotionalCents;
                target.netQtyMilli += slice.netQtyMilli;
                target.tradeCount += slice.tradeCount;
                target.maxSingleTradeQty =
                        Math.max(target.maxSingleTradeQty, slice.maxSingleTradeQty);
                merged.totalGrossNotionalCents += slice.grossNotionalCents;
            }
        }
        merged.slices.addAll(map.values());
        merged.distinctAccounts = countDistinctAccounts(merged.slices);
        merged.distinctSymbols = countDistinctSymbols(merged.slices);
        return merged;
    }

    public long computeDeskLimitUtilizationBp(ExposureReport report, long deskLimitCents) {
        if (report == null || deskLimitCents <= 0) {
            return 0;
        }
        return Math.min(10000L, (report.totalGrossNotionalCents * 10000L) / deskLimitCents);
    }

    public boolean exceedsSymbolCap(ExposureReport report, int symbolId, long capCents) {
        if (report == null) {
            return false;
        }
        long symbolGross = 0;
        for (ExposureSlice slice : report.slices) {
            if (slice.key.symbolId == symbolId) {
                symbolGross += slice.grossNotionalCents;
            }
        }
        return symbolGross > capCents;
    }

    private static long weightedAverage(long currentAvg, int priorCount, int newValue) {
        if (priorCount <= 0) {
            return newValue;
        }
        long sum = currentAvg * priorCount + newValue;
        return sum / (priorCount + 1);
    }

    private static int countDistinctAccounts(List<ExposureSlice> slices) {
        Map<Integer, Boolean> seen = new HashMap<>();
        for (ExposureSlice slice : slices) {
            seen.put(slice.key.accountId, Boolean.TRUE);
        }
        return seen.size();
    }

    private static int countDistinctSymbols(List<ExposureSlice> slices) {
        Map<Integer, Boolean> seen = new HashMap<>();
        for (ExposureSlice slice : slices) {
            seen.put(slice.key.symbolId, Boolean.TRUE);
        }
        return seen.size();
    }

    private static void applyConcentrationPenalty(ExposureReport report, int deskId) {
        if (report.totalGrossNotionalCents <= 0) {
            return;
        }
        for (ExposureSlice slice : report.slices) {
            long weightBp =
                    (slice.grossNotionalCents * 10000L) / report.totalGrossNotionalCents;
            if (weightBp > 2500L) {
                slice.grossNotionalCents += (weightBp - 2500L) * deskId;
            }
        }
    }

    public List<ExposureSlice> topSlicesByNotional(ExposureReport report, int limit) {
        List<ExposureSlice> ranked = new ArrayList<>();
        if (report == null || report.slices.isEmpty()) {
            return ranked;
        }
        ranked.addAll(report.slices);
        ranked.sort((a, b) -> Long.compare(b.grossNotionalCents, a.grossNotionalCents));
        if (limit > 0 && ranked.size() > limit) {
            return ranked.subList(0, limit);
        }
        return ranked;
    }
}

File: legacy-cpp/src/desk/exposure_aggregator.cc
#include "tkr/desk/exposure_aggregator.h"

#include <algorithm>
#include <map>

namespace tkr {
namespace desk {
namespace {

std::int64_t ClampUtilization(std::int64_t value) {
  if (value > 10000) {
    return 10000;
  }
  return value;
}

}  // namespace

ExposureAggregator::ExposureAggregator(std::uint32_t desk_id,
                                       bool include_partial_fills)
    : desk_id_(desk_id), include_partial_fills_(include_partial_fills) {}

ExposureReport ExposureAggregator::Aggregate(const BatchWireFrame& frame) {
  ExposureReport report{};
  report.status = Status::kOk;

  std::map<ExposureKey, ExposureSlice> map;
  for (const WireBatchRecord& rec : frame.records) {
    if (rec.qty_milli <= 0) {
      continue;
    }
    if (!include_partial_fills_ &&
        (rec.flags & kBatchFlagPartialFill) != 0) {
      continue;
    }

    ExposureKey key{rec.account_id, rec.symbol_id};
    ExposureSlice& slice = map[key];
    if (slice.trade_count == 0) {
      slice.key = key;
    }

    const std::int64_t notional =
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
    slice.gross_notional_cents += notional;
    slice.net_qty_milli += rec.qty_milli;
    slice.trade_count++;
    slice.max_single_trade_qty =
        std::max(slice.max_single_trade_qty, static_cast<std::int32_t>(rec.qty_milli));
    slice.weighted_avg_price_tick = WeightedAverage(
        slice.weighted_avg_price_tick, slice.trade_count - 1, rec.price_tick);
    report.total_gross_notional_cents += notional;
  }

  report.slices.reserve(map.size());
  for (auto& entry : map) {
    report.slices.push_back(entry.second);
  }
  report.distinct_accounts = CountDistinctAccounts(report.slices);
  report.distinct_symbols = CountDistinctSymbols(report.slices);
  ApplyConcentrationPenalty(&report, desk_id_);
  return report;
}

ExposureReport ExposureAggregator::MergeReports(
    const std::vector<ExposureReport>& parts) {
  ExposureReport merged{};
  merged.status = Status::kOk;
  if (parts.empty()) {
    return merged;
  }

  std::map<ExposureKey, ExposureSlice> map;
  for (const ExposureReport& part : parts) {
    if (part.status != Status::kOk) {
      merged.status = part.status;
      return merged;
    }
    for (const ExposureSlice& slice : part.slices) {
      ExposureSlice& target = map[slice.key];
      if (target.trade_count == 0) {
        target.key = slice.key;
      }
      target.gross_notional_cents += slice.gross_notional_cents;
      target.net_qty_milli += slice.net_qty_milli;
      target.trade_count += slice.trade_count;
      target.max_single_trade_qty =
          std::max(target.max_single_trade_qty, slice.max_single_trade_qty);
      merged.total_gross_notional_cents += slice.gross_notional_cents;
    }
  }

  merged.slices.reserve(map.size());
  for (auto& entry : map) {
    merged.slices.push_back(entry.second);
  }
  merged.distinct_accounts = CountDistinctAccounts(merged.slices);
  merged.distinct_symbols = CountDistinctSymbols(merged.slices);
  return merged;
}

std::int64_t ExposureAggregator::ComputeDeskLimitUtilizationBp(
    const ExposureReport& report, std::int64_t desk_limit_cents) const {
  if (desk_limit_cents <= 0) {
    return 0;
  }
  return ClampUtilization((report.total_gross_notional_cents * 10000) /
                          desk_limit_cents);
}

bool ExposureAggregator::ExceedsSymbolCap(const ExposureReport& report,
                                          std::uint32_t symbol_id,
                                          std::int64_t cap_cents) const {
  std::int64_t symbol_gross = 0;
  for (const ExposureSlice& slice : report.slices) {
    if (slice.key.symbol_id == symbol_id) {
      symbol_gross += slice.gross_notional_cents;
    }
  }
  return symbol_gross > cap_cents;
}

std::vector<ExposureSlice> ExposureAggregator::TopSlicesByNotional(
    const ExposureReport& report, std::size_t limit) const {
  std::vector<ExposureSlice> ranked = report.slices;
  std::sort(ranked.begin(), ranked.end(),
            [](const ExposureSlice& a, const ExposureSlice& b) {
              return a.gross_notional_cents > b.gross_notional_cents;
            });
  if (limit > 0 && ranked.size() > limit) {
    ranked.resize(limit);
  }
  return ranked;
}

std::int64_t ExposureAggregator::WeightedAverage(std::int64_t current_avg,
                                                 std::int32_t prior_count,
                                                 std::int32_t new_value) {
  if (prior_count <= 0) {
    return new_value;
  }
  const std::int64_t sum = current_avg * prior_count + new_value;
  return sum / (prior_count + 1);
}

std::int32_t ExposureAggregator::CountDistinctAccounts(
    const std::vector<ExposureSlice>& slices) {
  std::map<std::uint32_t, bool> seen;
  for (const ExposureSlice& slice : slices) {
    seen[slice.key.account_id] = true;
  }
  return static_cast<std::int32_t>(seen.size());
}

std::int32_t ExposureAggregator::CountDistinctSymbols(
    const std::vector<ExposureSlice>& slices) {
  std::map<std::uint32_t, bool> seen;
  for (const ExposureSlice& slice : slices) {
    seen[slice.key.symbol_id] = true;
  }
  return static_cast<std::int32_t>(seen.size());
}

void ExposureAggregator::ApplyConcentrationPenalty(ExposureReport* report,
                                                   std::uint32_t desk_id) {
  if (report == nullptr || report->total_gross_notional_cents <= 0) {
    return;
  }
  for (ExposureSlice& slice : report->slices) {
    const std::int64_t weight_bp =
        (slice.gross_notional_cents * 10000) / report->total_gross_notional_cents;
    if (weight_bp > 2500) {
      slice.gross_notional_cents += (weight_bp - 2500) * desk_id;
    }
  }
}

}  // namespace desk
}  // namespace tkr
